content/en/blog/_posts/Deploying-External-OpenStack-Cloud-Provider-With-Kubeadm.md

```diff
@@ -0,0 +1,742 @@
+
+## Install Docker CE.
+
+yum update && yum install docker-ce-18.06.2.ce
+
+## Create /etc/docker directory.
+
+mkdir /etc/docker
+
+# Setup daemon.
+
+cat > /etc/docker/daemon.json <<EOF
+{
+  "exec-opts": ["native.cgroupdriver=systemd"],
+  "log-driver": "json-file",
+  "log-opts": {
+    "max-size": "100m"
+  },
+  "storage-driver": "overlay2",
+  "storage-opts": [
+    "overlay2.override_kernel_check=true"
+  ]
+}
+EOF
+
+mkdir -p /etc/systemd/system/docker.service.d
+
+# Restart Docker
+systemctl daemon-reload
+systemctl restart docker
+systemctl enable docker
+```
+
+Install kubeadm following the steps in https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/
+
+```
+cat <<EOF > /etc/yum.repos.d/kubernetes.repo
+[kubernetes]
+name=Kubernetes
+baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
+enabled=1
+gpgcheck=1
+repo_gpgcheck=1
+gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
+EOF
+
+# Set SELinux in permissive mode (effectively disabling it)
+setenforce 0
+sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config
+
+yum install -y kubelet kubeadm kubectl --disableexcludes=kubernetes
+
+systemctl enable --now kubelet
+
+cat <<EOF >  /etc/sysctl.d/k8s.conf
+net.bridge.bridge-nf-call-ip6tables = 1
+net.bridge.bridge-nf-call-iptables = 1
+EOF
+sysctl --system
+
+# check if br_netfilter module is loaded
+lsmod | grep br_netfilter
+
+# if not, load it explicitly with 
+modprobe br_netfilter
+```
+
+The official document about how to create a single-master cluster can be found in https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/
+
+We'll largely follow that document but also add additional things for the cloud provider.
+To make things clearer we'll use a kubeadm-config.yml for the master.
+In this config we specify to use an external OpenStack cloud provider, and where to find its config.
+We also enable storage API in API server's runtime config so we can use OpenStack volumes as persistent volumes in Kubernets.
+
+```
+apiVersion: kubeadm.k8s.io/v1beta1
+kind: InitConfiguration
+nodeRegistration:
+  kubeletExtraArgs:
+    cloud-provider: "external"
+---
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+kubernetesVersion: "v1.15.1"
+apiServer:
+  extraArgs:
+    enable-admission-plugins: NodeRestriction
+    runtime-config: "storage.k8s.io/v1=true"
+controllerManager:
+  extraArgs:
+    external-cloud-volume-plugin: openstack
+  extraVolumes:
+  - name: "cloud-config"
+    hostPath: "/etc/kubernetes/cloud-config"
+    mountPath: "/etc/kubernetes/cloud-config"
+    readOnly: true
+    pathType: File
+networking:
+  serviceSubnet: "10.96.0.0/12"
+  podSubnet: "10.224.0.0/16"
+  dnsDomain: "cluster.local"
+```
+
+Now we'll create cloud config, /etc/kubernetes/cloud-config, for OpenStack. 
```

Comment:
```suggestion
Now we'll create cloud config, `/etc/kubernetes/cloud-config`, for OpenStack. 
```
File paths should use code-styling. 
See the style guide for more info: https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-filenames-directories-and-paths